### PR TITLE
Namespace restricted mode

### PR DESCRIPTION
**Pull request by nimrod-up9**
Mizu's default behavior is to create a new `mizu` namespace and to install itself in that namespace.
This PR adds a `tap.agent: <NAMESPACE>` config that instead installs Mizu in an existing namespace.
Note that in this mode IP resolving is restricted to <NAMESPACE> (though a user may tap all namespace if they have sufficient permissions).

Example:
`mizu tap <PATTERN> --set agent.namespace=<USER_NAMESPACE> -n <USER_NAMESPACE>`
will only create mizu resources and tap pods in `<USER_NAMESPACE>`. This only requires the user to have permissions to this particular namespace.

### CODE REVIEW

**Comment by nimrod-up9** [comment]
Hi @ksudhir007, I would like to get your opinion on this feature. Does it fit your workflow of splitting a cluster between several users by namespace?

File: cli/cmd/tap.go [hunk 1]
```diff
@@ -100,6 +102,7 @@ func init() {
 	tapCmd.Flags().BoolVarP(&mizuTapOptions.AllNamespaces, "all-namespaces", "A", false, "Tap all namespaces")
 	tapCmd.Flags().StringVarP(&mizuTapOptions.KubeConfigPath, "kube-config", "k", "", "Path to kube-config file")
 	tapCmd.Flags().StringVarP(&mizuTapOptions.MizuImage, "mizu-image", "", fmt.Sprintf("gcr.io/up9-docker-hub/mizu/%s:%s", mizu.Branch, mizu.SemVer), "Custom image for mizu API server")
+	tapCmd.Flags().StringVarP(&mizuTapOptions.MizuNamespace, "mizu-namespace", "", "", "A existing namespace in which to install mizu resources. If not passed, create a temporary namespace \"mizu\" for this purpose.")
```
Comment (IgorGov) [inline cli/cmd/tap.go:105]: "An existing..."

File: cli/cmd/view.go [hunk 2]
```diff
@@ -32,4 +33,5 @@ func init() {
 
 	viewCmd.Flags().Uint16VarP(&mizuViewOptions.GuiPort, "gui-port", "p", 8899, "Provide a custom port for the web interface webserver")
 	viewCmd.Flags().StringVarP(&mizuViewOptions.KubeConfigPath, "kube-config", "k", "", "Path to kube-config file")
+	viewCmd.Flags().StringVarP(&mizuViewOptions.MizuNamespace, "mizu-namespace", "", "", "A existing namespace in which to install mizu resources. If not passed, create a temporary namespace \"mizu\" for this purpose.")
```
Comment (IgorGov) [inline cli/cmd/view.go:36]: An existing..

File: examples/roles/permissions-ns-without-ip-resolution.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# This example shows the roles required for a user to be able to use Mizu in all namespaces.
```
Comment (IgorGov) [inline examples/roles/permissions-ns-without-ip-resolution.yaml:1]: should the description include "the without ip resolution"?

File: examples/roles/permissions-ns.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# This example shows the roles required for a user to be able to use Mizu in all namespaces.
```
Comment (IgorGov) [inline examples/roles/permissions-ns.yaml:1]: is it the correct description?

File: Makefile [hunk 5]
```diff
@@ -32,13 +32,8 @@ agent: ## build mizuagent server
...
 	@ls -l agent/build
 
-#tap: ## build tap binary
-#	@(cd tap; go build -o build/tap ./src)
-#	@ls -l tap/build
-
 docker: ## build Docker image 
-	@(echo "building docker image" )
-	./build-push-featurebranch.sh
+	$(MAKE) push-docker
```
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply Makefile:36]: I agree. We need to improve the development workflow.

File: cli/cmd/tap.go (hunk 1 above)
Comment (nimrod-up9) [reply cli/cmd/tap.go:105]: done

File: cli/cmd/view.go (hunk 2 above)
Comment (nimrod-up9) [reply cli/cmd/view.go:36]: done

File: examples/roles/permissions-ns-without-ip-resolution.yaml (hunk 3 above)
Comment (nimrod-up9) [reply examples/roles/permissions-ns-without-ip-resolution.yaml:1]: fixed

File: examples/roles/permissions-ns.yaml (hunk 4 above)
Comment (nimrod-up9) [reply examples/roles/permissions-ns.yaml:1]: fixed

**Comment by ksudhir007** [comment]
> Hi @ksudhir007, I would like to get your opinion on this feature. Does it fit your workflow of sharing a cluster between several users by namespace?

Thank you! this should work. I can try this out once its ready.

File: agent/pkg/resolver/resolver.go [hunk 6]
```diff
@@ -17,21 +17,37 @@ const (
...
+	clientSet    *kubernetes.Clientset
+	nameMap      map[string]string
+	serviceMap   map[string]string
+	isStarted    bool
+	errOut       chan error
+	namespace    string
 }
 
-func (resolver *Resolver) Start(ctx context.Context) {
+func (resolver *Resolver) Start(ctx context.Context, namespace string) {
```
Comment (nimrod-up9) [inline agent/pkg/resolver/resolver.go:42]: Why pass namespace if there is resolver.namespace?

File: cli/cmd/tapRunner.go [hunk 7]
```diff
@@ -19,46 +11,81 @@ import (
...
 	cleanupTimeout     = time.Minute
+	updateTappersDelay = 5 * time.Second
 )
 
-var currentlyTappedPods []core.Pod
+type tapCmdBL struct {
+	apiServerService         *core.Service
+	currentlyTappedPods      []core.Pod
+	flags                    *MizuTapOptions
+	mizuServiceAccountExists bool
```
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply cli/cmd/tapRunner.go:35]: Correction: These are two different things.
`apiServerService` is the API service.
`mizuServiceAccountExists` refers to the service **account**.

File: agent/pkg/resolver/resolver.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply agent/pkg/resolver/resolver.go:30]: done

[the remark this answers is not in the document]
Comment (IgorGov) [reply agent/pkg/resolver/resolver.go:30]: i can still see the "namespace" var?

File: cli/mizu/configStruct.go [hunk 8]
```diff
@@ -7,14 +7,31 @@ import (
...
+	Telemetry     bool                        `yaml:"telemetry" default:"true"`
 }
 
 func (config *ConfigStruct) SetDefaults() {
 	config.MizuImage = fmt.Sprintf("gcr.io/up9-docker-hub/mizu/%s:%s", Branch, SemVer)
 }
+
+func (config *ConfigStruct) ResourcesNamespace() string {
+	if config.MizuNamespace == "" {
+		return ResourcesDefaultNamespace
```
Comment (IgorGov) [inline cli/mizu/configStruct.go:25]: why not setting "mizu" as default namespace?

File: cli/mizu/configStruct.go [hunk 9]
```diff
@@ -7,14 +7,31 @@ import (
...
+	}
+
+	return config.MizuNamespace
+}
+
+func (config *ConfigStruct) IsOwnNamespace() bool {
+	if config.MizuNamespace == "" {
```
Comment (IgorGov) [inline cli/mizu/configStruct.go:32]: here you can check config.MizuNamespace == ResourcesDefaultNamespace

File: cli/cmd/tapRunner.go [hunk 10]
```diff
@@ -131,50 +136,57 @@ func createMizuResources(ctx context.Context, kubernetesProvider *kubernetes.Pro
...
 	return nil
 }
 
 func createMizuConfigmap(ctx context.Context, kubernetesProvider *kubernetes.Provider, data string) error {
-	err := kubernetesProvider.ApplyConfigMap(ctx, mizu.ResourcesNamespace, mizu.ConfigMapName, data)
-	if err != nil {
-		fmt.Printf("Error creating mizu configmap: %v\n", err)
-	}
-	return nil
+	err := kubernetesProvider.CreateConfigMap(ctx, mizu.Config.ResourcesNamespace(), mizu.ConfigMapName, data)
```
Comment (IgorGov) [inline cli/cmd/tapRunner.go:149]: do we really need this wrapper?

File: cli/cmd/tapRunner.go [hunk 11]
```diff
@@ -131,50 +136,57 @@ func createMizuResources(ctx context.Context, kubernetesProvider *kubernetes.Pro
...
 
 func createMizuNamespace(ctx context.Context, kubernetesProvider *kubernetes.Provider) error {
-	_, err := kubernetesProvider.CreateNamespace(ctx, mizu.ResourcesNamespace)
-	if err != nil {
-		mizu.Log.Infof("Error creating Namespace %s: %v", mizu.ResourcesNamespace, err)
-		return err
-	}
-	mizu.Log.Debugf("Successfully creating Namespace %s", mizu.ResourcesNamespace)
-	return nil
+	_, err := kubernetesProvider.CreateNamespace(ctx, mizu.Config.ResourcesNamespace())
```
Comment (IgorGov) [inline cli/cmd/tapRunner.go:154]: redundant wrapper ?

File: cli/cmd/tapRunner.go [hunk 12]
```diff
@@ -241,31 +249,73 @@ func cleanUpMizuResources(kubernetesProvider *kubernetes.Provider) {
 	removalCtx, cancel := context.WithTimeout(context.Background(), cleanupTimeout)
 	defer cancel()
 
-	if err := kubernetesProvider.RemoveNamespace(removalCtx, mizu.ResourcesNamespace); err != nil {
-		mizu.Log.Infof("Error removing Namespace %s: %s (%v,%+v)", mizu.ResourcesNamespace, err, err, err)
-		return
+	if mizu.Config.IsOwnNamespace() {
+		if err := kubernetesProvider.RemoveNamespace(removalCtx, mizu.Config.ResourcesNamespace()); err != nil {
```
Comment (IgorGov) [inline cli/cmd/tapRunner.go:253]: consider moving this functionality to function

**Review by IgorGov — APPROVED** [review]

**Pull request merged**